File: packaging/scripts/prepare.py
#!/bin/sh
# Compatibility shim for legacy Python prepare.py invocation.
#
# Python usage:  prepare.py -uvtd
# Go equivalent: ztc prepare --all
#
# Short flag mapping:
#   -u (utils)     -> -u (accepted, no-op in Go)
#   -v (vhosts)    -> -V (virtual-hosts; can't reuse -v, it's global --verbose)
#   -t (template)  -> -t (templates)
#   -d (dashboard) -> -d (dashboard)
#
# Long flag mapping:
#   --vhosts    -> --virtual-hosts
#   --template  -> --templates
#
# No-args behavior:
#   Legacy: shows help and exits (no mutation).
#   Go:     defaults to --all (creates all objects).
#   This shim matches legacy behavior — no args prints a warning and exits.

# If no arguments given, match legacy behavior: show help, don't mutate.
if [ $# -eq 0 ]; then
    echo "WARNING: prepare.py called with no arguments." >&2
    echo "Legacy behavior: show help and exit (no changes made)." >&2
    echo "" >&2
    echo "To create all Zabbix objects, run:" >&2
    echo "  ztc prepare --all" >&2
    echo "  ztc prepare --all --force   # after upgrade from Python version" >&2
    echo "" >&2
    echo "Available flags:" >&2
    echo "  -u  check utility paths (no-op in Go)" >&2
    echo "  -v  create virtual hosts" >&2
    echo "  -t  create/update templates" >&2
    echo "  -d  create dashboard" >&2
    exit 0
fi

# Rewrite args: short flags and long flags.
ARGS=""
for arg in "$@"; do
    case "$arg" in
        --vhosts)    ARGS="$ARGS --virtual-hosts" ;;
        --template)  ARGS="$ARGS --templates" ;;
        --*)         ARGS="$ARGS $arg" ;;          # pass other long flags through
        -*)          ARGS="$ARGS $(echo "$arg" | sed 's/v/V/g')" ;;  # -uvtd -> -uVtd
        *)           ARGS="$ARGS $arg" ;;          # positional args
    esac
done

# shellcheck disable=SC2086
exec /usr/bin/ztc prepare $ARGS

File: cmd/prepare.go
package cmd

import (
	"context"
	"fmt"

	"github.com/spf13/cobra"
)

var (
	prepareTemplates    bool
	prepareVirtualHosts bool
	prepareDashboard    bool
	prepareActions      bool
	prepareAll          bool
	prepareForce        bool
	prepareUtils        bool // hidden: Python -u compat (no-op in Go)
)

var prepareCmd = &cobra.Command{
	Use:   "prepare",
	Short: "Prepare Zabbix objects for vulnerability monitoring",
	Long: `Create and configure Zabbix objects required for vulnerability monitoring.

This command can create:
- OS-Report template for package collection (-t)
- Virtual hosts for aggregated vulnerability data (-V)
- Dashboards for vulnerability visualization (-d)
- Actions: checked but require manual configuration in the Zabbix UI (-A)

When upgrading from the Python version, run with --force to recreate
templates and discovery rules with the new key schema.

NOTE: This command does not require a Vulners API key.`,
	RunE: func(cmd *cobra.Command, args []string) error {
		log := GetLogger()
		cfg := GetConfig()

		log.Info("Preparing Zabbix objects...")

		client, err := initZabbixClient(cfg, log)
		if err != nil {
			return fmt.Errorf("failed to connect to Zabbix: %w", err)
		}
		defer func() { _ = client.Close() }()

		// Default to all when no specific flags are given.
		// This matches the typical usage (Python: prepare.py -uvtd)
		// and avoids a silent no-op when migration docs say "run ztc prepare".
		noFlagsSet := !prepareAll && !prepareTemplates && !prepareVirtualHosts && !prepareDashboard && !prepareActions
		if noFlagsSet {
			log.Warn("No flags specified, defaulting to --all (create all Zabbix objects)")
		}
		if prepareAll || noFlagsSet {
			prepareTemplates = true
			prepareVirtualHosts = true
			prepareDashboard = true
			prepareActions = true
		}

		ctx := context.Background()

		if prepareForce {
			log.Warn("Force mode enabled — existing objects will be recreated")
		}

		if prepareTemplates {
			log.Info("Creating/updating OS-Report template...")
			if err := client.EnsureOSReportTemplateCtx(ctx, prepareForce); err != nil {
				return fmt.Errorf("failed to create template: %w", err)
			}
			log.Info("OS-Report template ready")
		}

		if prepareVirtualHosts {
			log.Info("Creating virtual hosts...")
			if err := client.EnsureVirtualHostsCtx(ctx, prepareForce); err != nil {
				return fmt.Errorf("failed to create virtual hosts: %w", err)
			}
			log.Info("Virtual hosts ready")
		}

		if prepareDashboard {
			log.Info("Creating dashboard...")
			if err := client.EnsureDashboardCtx(ctx, prepareForce); err != nil {
				return fmt.Errorf("failed to create dashboard: %w", err)
			}
			log.Info("Dashboard ready")
		}

		if prepareActions {
			log.Info("Checking actions...")
			if err := client.EnsureActionsCtx(ctx); err != nil {
				return fmt.Errorf("failed to check actions: %w", err)
			}
		}

		log.Info("Zabbix preparation complete")
		return nil
	},
}

func init() {
	prepareCmd.Flags().BoolVarP(&prepareAll, "all", "a", false, "create all Zabbix objects (default when no flags given)")
	prepareCmd.Flags().BoolVarP(&prepareTemplates, "templates", "t", false, "create/update OS-Report template")
	prepareCmd.Flags().BoolVarP(&prepareVirtualHosts, "virtual-hosts", "V", false, "create virtual hosts")
	prepareCmd.Flags().BoolVarP(&prepareDashboard, "dashboard", "d", false, "create dashboard")
	prepareCmd.Flags().BoolVarP(&prepareActions, "actions", "A", false, "check if actions exist (manual Zabbix UI setup required)")
	prepareCmd.Flags().BoolVarP(&prepareForce, "force", "f", false, "recreate existing objects (use after upgrade to fix key schema changes)")

	// Hidden Python-compat flags so "prepare -uvtd" doesn't fail.
	// -u (--utils): Python checked zabbix-sender/get paths; Go does this implicitly.
	prepareCmd.Flags().BoolVarP(&prepareUtils, "utils", "u", false, "check utility paths (accepted for Python compat, no-op)")
	_ = prepareCmd.Flags().MarkHidden("utils")

	rootCmd.AddCommand(prepareCmd)
}
